### PR TITLE
Ensure `npm bin` is in $PATH when installing globally

### LINKED ISSUE #6993: check $PATH when installing with '-g'
When installing with '-g' NPM should check if the 'global' bin folder is in the $PATH, and propose a fix to the user. Given that NPM and node can be installed in so many different ways, it can definitely help newbies. See https://github.com/mcollina/mosca/issues/206 for more details.

### PR DESCRIPTION
**Pull request by ariporad**
Fixes #6993.

### CODE REVIEW

**Comment by ariporad** [comment]
Also, the travis build is failing, and I'm 99.99% sure it's not what I did. Could someone look at it for me? Thanks!

**Comment by iarna** [comment]
@ariporad It's not, https://github.com/npm/npm/pull/10192 is the fix for it.

**Comment by iarna** [comment]
The two test files in https://github.com/ariporad/npm/commit/67c1f193e805a66602447da694b09983d76ab4ae shouldn't be included. Also it looks like you had $PATH expand when you ran commit there?

**Comment by ariporad** [comment]
@iarna: Ok, fixed $PATH, removed `/test/tap/dedupe`. Should be ready to go.

One other thing: I realized that with this change, running `npm bin -g` won't warn if the global npm bin isn't in $PATH. I don't know if this is a problem, but if it is, I'll fix it.

**Comment by othiym23** [comment]
The code is looking good, and now we need to add a tests before we merge and release this. Do you want to take a pass at it, and if so, how can we help you get started writing an npm test? (They're a little tricky.)

**Comment by ariporad** [comment]
@othiym23: Sure, I'll try it. Where would I start?

**Comment by ariporad** [comment]
@iarna, @othiym23: bump: I'd love to try to write some tests for it, where would I start?

File: lib/bin.js [hunk 1]
```diff
@@ -11,12 +11,21 @@ function bin (args, silent, cb) {
...
-  var PATH = osenv.path()
 
   if (!silent) console.log(b)
   process.nextTick(cb.bind(this, null, b))
+}
 
-  if (npm.config.get('global') && PATH.indexOf(b) === -1) {
-    npm.config.get('logstream').write('(not in PATH env variable)\n')
-  }
+bin.ensureInPath = function ensureInPath (cb) {
```
Comment (iarna) [inline lib/bin.js]: So the first thing I'd do is write a functional test for this part. To do that you'll want to be able to control all of the variables that go into this:
```
var requireInject = require('require-inject')
var binPath = '/usr/local/bin/npm' // or whatever
var envPATH = '/usr/bin:/bin:/usr/local/bin'
var isGlobal = true
var bin = requireInject('../../lib/bin.js', {
  npm: {
    bin: binPath,
    config: {
      get: function (key) {
          if (key !== 'global') throw new Error('unexpected config key: ' + key)
          return isGlobal
      }
    }
  },
  osenv: {
    path: function () {
      return envPATH
    }
  }
})

File: lib/install.js [hunk 2]
```diff
@@ -289,7 +290,12 @@ Installer.prototype.run = function (cb) {
         log.warn('error', postInstallEr.message)
         log.verbose('error', postInstallEr.stack)
       }
-      cb(installEr || postInstallEr, self.getInstalledModules(), self.idealTree)
+      ensureInPath(function isInPath (isInPathEr, isInPath) {
```
Comment (iarna) [inline lib/install.js]: I think this test should probably go in `/lib/install/validate-tree` but I'll grant that we're not exactly validating the tree here, this is more about validating your install environment, so I don't hate having it in `install.js` either. But regardless, it shouldn't be calling `log.warn`, instead it should be constructing an error object and pushing it on to `idealTree.warnings`, which will be printed out after the installation summary.

**Comment by iarna** [comment]
Hey @ariporad I finally got a chance to give some hints on writing a test for this. I also had one comment on your fix here.

Tests go in `test/tap/name-of-test.js` – you can look at any of the others there to get a sense of the boiler plate, but it boils down to:
```
var test = require('tap').test
var common = require('../common-npm.js')

test('name-of-group-of-tests', function (t) {
  t.is(actual, expected, 'description of what this is testing')
  t.done()
})
```

**Comment by iarna** [comment]
If you've got further questions please feel free to ask here, (or if you want a faster turn around, [DM me on Twitter](https://twitter.com/ReBeccaOrg))

**Comment by iarna** [comment]
We're closing this pull request as it has gone sixty days without activity.  In our experience pull requests that go this long are unlikely to land.  If you have updates to it, please feel free to open a new pull request.

For more information about our new issue aging policies and why we've instituted them please see our [blog post](http://blog.npmjs.org/post/161832149430/npm-the-npm-github-issue-tracker-and-you).

**Pull request closed without merging**